tests/integrationv2/test_npn.py

```diff
@@ -0,0 +1,204 @@
+
+    expected_protocol = 'http/1.1'
+
+    for results in server.get_results():
+        results.assert_success()
+        assert to_bytes(OPENSSL_SERVER_NPN_MARKER + expected_protocol) in results.stdout
+
+    for results in s2n_client.get_results():
+        results.assert_success()
+        assert to_bytes(S2N_NPN_MARKER) in results.stdout
+        assert to_bytes(S2N_APPLICATION_MARKER + expected_protocol) in results.stdout
+
+"""
+The s2n-tls client chooses a server-preferred protocol.
+"""
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", ALL_TEST_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", MINIMAL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", TLS_PROTOCOLS, ids=get_parameter_name)
+@pytest.mark.parametrize("provider", [OpenSSL], ids=get_parameter_name)
+def test_s2n_client_npn_server_preference(managed_process, cipher, curve, certificate, protocol, provider):
+    s2n_client, server = s2n_client_npn_handshake(managed_process, cipher, curve, certificate, protocol, provider, \
+        server_list=PROTOCOL_LIST_ALT_ORDER)
+
+    expected_protocol = 'h2'
+
+    for results in server.get_results():
+        results.assert_success()
+        assert to_bytes(OPENSSL_SERVER_NPN_MARKER + expected_protocol) in results.stdout
+
+    for results in s2n_client.get_results():
+        results.assert_success()
+        assert to_bytes(S2N_NPN_MARKER) in results.stdout
+        assert to_bytes(S2N_APPLICATION_MARKER + expected_protocol) in results.stdout
+
+"""
+The s2n-tls client chooses its preferred protocol since there is no overlap.
+"""
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", ALL_TEST_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("curve", ALL_TEST_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", MINIMAL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", TLS_PROTOCOLS, ids=get_parameter_name)
+@pytest.mark.parametrize("provider", [OpenSSL], ids=get_parameter_name)
+def test_s2n_client_npn_no_overlap(managed_process, cipher, curve, certificate, protocol, provider):
+    s2n_client, server = s2n_client_npn_handshake(managed_process, cipher, curve, certificate, protocol, provider, \
+        server_list=PROTOCOL_LIST_NO_OVERLAP)
+
+    expected_protocol = 'http/1.1'
+
+    for results in server.get_results():
+        results.assert_success()
+        assert to_bytes(OPENSSL_SERVER_NPN_MARKER + expected_protocol) in results.stdout
+
+    for results in s2n_client.get_results():
+        results.assert_success()
+        assert to_bytes(S2N_NPN_MARKER) in results.stdout
+        assert to_bytes(S2N_APPLICATION_MARKER + expected_protocol) in results.stdout
+
+
+def s2n_server_npn_handshake(managed_process, cipher, curve, certificate, protocol, provider, server_list):
```

Comment:
This looks identical to `s2n_client_npn_handshake`, does these need to be two separate functions?

Reply:
Since we're switching from s2n being a server to a client you'd have to add an if-statement to handle that and I just wasn't interested in adding another parameter to the function. I did consider this though, and decided it wasn't going to make the code any cleaner.